### ISSUE TITLE
is there any fast method to add regularization factor？

### ISSUE THREAD

**Issue by xyzhou-puck** [torch/torch7#302]
Hi guys,

Just wondering if there is any fast method to add regularization factor before update the parameters, I tried to use getParameters() function to get a pointer of the current parameters and gradients and then I use add(_sim_parameters:clone():mul(l2_norm)) to change the gradients for regularization, and finally do updates, however the method above is pretty slow when I runs on GPU, could you please give me some advices?

Thanks,
Xiangyang

**Comment by hughperkins** [comment]
getParameters is mega slow.  It calls `get` for every point in the tensor.  What I did was read the parameters directly from each module, according to the type of module, using an if:

```
if torch.type(module) == 'nn.Linear' then
   parameters = module.weight -- and you need to get the bias too of course
-- elseif ... etc ...
end
```

It's not the best solution, but it's an easy short-term solution.

**Comment by adamlerer** [comment]
@hughperkins you're right that `getParameters` is extremely slow because it flattens all your parameters into one contiguous block and fixes up all the model parameters to point to that (it should really be called `flattenParameters`). But you should only have to call `getParameters` once at the beginning of training, e.g.

```lua
parameters, gradParameters = model:getParameters()
for i = 1, nBatch do
  train(model)
  parameters:add(lr, gradParameters)
  regularize(parameters)
end
```

You can also get the model parameters without "flattening" them (as a table of tensors) by calling `parameters()`.

```lua
parameters, gradParameters = model:parameters()
for i = 1, nBatch do
  train(model)
  for j = 1, #parameters do
    parameters[j]:add(lr, gradParameters[j])
    ... regularize ...
  end 
end

**Comment by hughperkins** [comment]
ah, good info.  thanks!

**Comment by xyzhou-puck** [comment]
thanks guys, I just found another way for regularization, so never need the
straightforward method any more, thanks.
